### PR TITLE
`AppTemplate::apply_batch()`  refactoring

### LINKED ISSUE #237: Make the state machine in sov-app-template explicit
Currently, the `apply_blob` function has an implicit state machine. At the start of each blob, we make the `batch_workspace` revertable, and then gradually begin applying transactions to it. Since the state machine is implicit, the type checker does not enforce that we leave the batch workspace in a coherent state - so it's easy to forget to commit or revert changes on unhappy code paths. We should make the state machine explicit, so that invalid transitions are rejected at compile time.

### PR DESCRIPTION
**Pull request by bkolad**
# Description
This PR introduces the following changes:

1. The  `CommittedWorkingSet` has been changed to `StateCheckpoint`.
2. The `AppTemplate` has been moved to its own file.
3. Some helper methods have been added to `AppTemplate`, and `apply_batch()` has been made more modular.
The helper methods now return the `Result<_, ApplyBatchError>` type, which allows us to use the `?` operator and make the shortcutting logic in `apply_batch()` more readable.

## Linked Issues
- Fixes #237, #388

## Testing
No impact on tests.
## Docs
Relevant docs updated

### CODE REVIEW

File: module-system/sov-modules-stf-template/clippy.toml [hunk 1]
```diff
@@ -0,0 +1 @@
+type-complexity-threshold  = 300
```
Comment (bkolad) [inline module-system/sov-modules-stf-template/clippy.toml:1]: `clippy` was not happy with this signature (return type):
```

    fn execute_txs(
        &mut self,
        mut batch_workspace: WorkingSet<C::Storage>,
        txs: Vec<TransactionAndRawHash<C>>,
        batch_data_and_hash: &BatchDataAndHash,
    ) -> Result<(WorkingSet<C::Storage>, Vec<TransactionReceipt<TxEffect>>)> {
...
}
```
But I don't thing it is that bad so I increased the `type-complexity-threshold`

File: module-system/sov-modules-stf-template/src/app_template.rs [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+type Result<T> = std::result::Result<T, ApplyBatchError>;
+
+pub struct AppTemplate<C: Context, RT, Vm> {
+    pub current_storage: C::Storage,
+    pub runtime: RT,
+    pub(crate) checkpoint: Option<StateCheckpoint<C::Storage>>,
+    phantom_vm: PhantomData<Vm>,
+}
+
+pub(crate) enum ApplyBatchError {
```
Comment (citizen-stig) [inline module-system/sov-modules-stf-template/src/app_template.rs:28]: Could you please add public doc comment explaining what kind of hash is in both cases?

File: module-system/sov-modules-stf-template/src/app_template.rs [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+                    let ctx = C::new(sender_address.clone());
+                    let tx_result = self.runtime.dispatch_call(msg, &mut batch_workspace, &ctx);
+
+                    self.runtime
+                        .post_dispatch_tx_hook(&tx.tx, &mut batch_workspace)
+                        .expect("Impossible happened: error in post_dispatch_tx_hook");
+
+                    let tx_effect = match tx_result {
+                        Ok(_) => TxEffect::Successful,
+                        Err(_e) => {
```
Comment (citizen-stig) [inline module-system/sov-modules-stf-template/src/app_template.rs:175]: Maybe let's log the error?

File: module-system/sov-modules-stf-template/src/app_template.rs (hunk 2 above)
Comment (bkolad) [reply module-system/sov-modules-stf-template/src/app_template.rs:28]: added

File: module-system/sov-modules-stf-template/src/app_template.rs (hunk 3 above)
Comment (bkolad) [reply module-system/sov-modules-stf-template/src/app_template.rs]: I put `debug` it is ok for a single tx to fail here.

**Review by preston-evans98 — APPROVED** [review]

**Pull request merged**